src/Moq/Mock`1.cs

```diff
@@ -633,12 +633,19 @@ public Mock<T> SetupProperty<TProperty>(Expression<Func<T, TProperty>> property,
 			Guard.NotNull(property, nameof(property));
 
 			var pi = property.ToPropertyInfo();
-			Guard.CanRead(pi);
-			Guard.CanWrite(pi);
 
-			TProperty value = initialValue;
-			this.SetupGet(property).Returns(() => value);
-			Mock.SetupSet(this, property.AssignItIsAny(), condition: null).SetCallbackBehavior(new Action<TProperty>(p => value = p));
+			if (!pi.CanRead(out var getter))
+			{
+				Guard.CanRead(pi);
+			}
+
+			if (!pi.CanWrite(out var setter))
+			{
+				Guard.CanWrite(pi);
+			}
```

Comment:
This isn't pretty. Seems like we're missing a method that acts as both guard and for accessor(s) retrieval.